atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java

```diff
@@ -750,20 +769,152 @@ public void partiallyFilledRowsShouldBeVisible() {
+            assertThat(e.getMessage(), containsString("Condition failed"));
+        }
+    }
+
+    @Test
+    public void readTransactionSucceedsIfConditionSucceeds() {
+        serializableTxManager.runTaskReadOnlyWithCondition(PreCommitCondition.NO_OP,
+                (tx, condition) -> tx.get(TABLE, ImmutableSet.of(TEST_CELL)));
+    }
+
+    @Test
+    public void readTransactionFailsIfConditionFails() {
+        try {
+            serializableTxManager.runTaskReadOnlyWithCondition(ALWAYS_FAILS_CONDITION,
+                    (tx, condition) -> tx.get(TABLE, ImmutableSet.of(TEST_CELL)));
+            fail();
+        } catch (TransactionFailedRetriableException e) {
+            assertThat(e.getMessage(), containsString("Condition failed"));
+        }
+    }
+
+    @Test
+    public void cleanupPreCommitConditionsOnSuccess() {
+        MutableLong counter = new MutableLong(0L);
+        PreCommitCondition succeedsCondition = new PreCommitCondition() {
+            @Override
+            public void throwIfConditionInvalid(long timestamp) {}
+
+            @Override
+            public void cleanup() {
+                counter.increment();
+            }
+        };
+
+        serializableTxManager.runTaskWithConditionThrowOnConflict(succeedsCondition, (tx, condition) -> {
+            tx.put(TABLE, ImmutableMap.of(TEST_CELL, PtBytes.toBytes("value")));
+            return null;
+        });
+        assertThat(counter.intValue(), is(1));
+
+        serializableTxManager.runTaskReadOnlyWithCondition(succeedsCondition,
+                (tx, condition) -> tx.get(TABLE, ImmutableSet.of(TEST_CELL)));
+        assertThat(counter.intValue(), is(2));
+    }
+
+    @Test
+    public void cleanupPreCommitConditionsOnFailure() {
+        MutableLong counter = new MutableLong(0L);
+        PreCommitCondition failsCondition = new PreCommitCondition() {
+            @Override
+            public void throwIfConditionInvalid(long timestamp) {
+                throw new TransactionFailedRetriableException("Condition failed");
+            }
+
+            @Override
+            public void cleanup() {
+                counter.increment();
+            }
+        };
+
+        try {
+            serializableTxManager.runTaskWithConditionThrowOnConflict(failsCondition, (tx, condition) -> {
+                tx.put(TABLE, ImmutableMap.of(TEST_CELL, PtBytes.toBytes("value")));
+                return null;
+            });
+            fail();
+        } catch (TransactionFailedRetriableException e) {
+            // expected
+        }
+        assertThat(counter.intValue(), is(1));
+
+        try {
+            serializableTxManager.runTaskReadOnlyWithCondition(failsCondition,
+                    (tx, condition) -> tx.get(TABLE, ImmutableSet.of(TEST_CELL)));
+            fail();
+        } catch (TransactionFailedRetriableException e) {
+            // expected
+        }
+        assertThat(counter.intValue(), is(2));
+    }
```

Comment:
nit: Seems like this could be two tests (read-only/general case), as well as the one above?

Reply:
I'm not sure it's worth it. There's either code duplication of the conditions or I move them up to non-static class variables and bloat the setup. I think this is a bit cleaner despite combining test cases.